modyn/supervisor/internal/triggers/ensembletrigger.py

```diff
@@ -0,0 +1,230 @@
+                # non triggering aggregation result
+
+                break
+
+            processing_head = next_trigger_idx
+
+            # reset state for next detection pass;
+            self._reset_subtrigger_decision_cache(subtrigger_decision_cache)
+            self._update_outdated_next_trigger_indexes(
+                processing_head,
+                subtrigger_generators,
+                next_subtrigger_index_cache,
+            )
+
+            yield next_trigger_idx
+
+        self.last_inform_decisions = subtrigger_decision_cache
+
+    def inform_new_model(self, most_recent_model_id: int) -> None:
+        for trigger in self.subtriggers.values():
+            trigger.inform_new_model(most_recent_model_id)
+
+    # --------------------------------------------------- Internal --------------------------------------------------- #
+
+    def _create_subtriggers(self, config: EnsembleTriggerConfig) -> dict[str, Trigger]:
+        subtriggers: dict[str, Trigger] = {}
+        for trigger_id, trigger_config in config.subtriggers.items():
+            if isinstance(trigger_config, TimeTriggerConfig):
+                subtriggers[trigger_id] = TimeTrigger(trigger_config)
+            elif isinstance(trigger_config, DataAmountTriggerConfig):
+                subtriggers[trigger_id] = DataAmountTrigger(trigger_config)
+            elif isinstance(trigger_config, DataDriftTriggerConfig):
+                subtriggers[trigger_id] = DataDriftTrigger(trigger_config)
+            elif isinstance(trigger_config, EnsembleTriggerConfig):
+                raise NotImplementedError("Nested ensemble triggers are not supported.")
+            else:
+                raise ValueError(f"Unknown trigger config: {trigger_config}")
+        return subtriggers
+
+    def _reset_subtrigger_decision_cache(self, subtrigger_decision_cache: dict[str, bool]) -> None:
+        subtrigger_decision_cache.clear()
+        subtrigger_decision_cache.update({trigger_name: False for trigger_name in self.subtriggers})
+
+    def _update_outdated_next_trigger_indexes(
+        self,
+        processing_head: int,
+        subtrigger_generators: dict[str, Generator[int, None, None]],
+        next_subtrigger_index_cache: dict[str, int | None],
+    ) -> None:
+        """Updates the next trigger indexes in the cache to be in the future
+        w.r.t. the current trigger index.
+
+        Supposed to be called after a new trigger index was found or at the beginning of the inform method to
+        sync. with the new data batch.
+
+        Note: Arguments are modified in place.
+        """
+        for trigger_name, generator in subtrigger_generators.items():
+            # generator is not exhausted
+            if next_subtrigger_index_cache[trigger_name] is not None:
+                try:
+                    while cast(dict[str, int], next_subtrigger_index_cache)[trigger_name] <= processing_head:
```

Comment:
maybe you can remove the casts by instead of using None use -2 (or two variables like NO_TRIGGER_YET = -1 and EMPTY_GENERATOR = -2)?

Reply:
Yep, `-2` sounds good. Mypy is simply too stupid to derive the strongest postcondition for the guard clauses.